crates/core/src/host_component.rs

```diff
@@ -0,0 +1,217 @@
+impl<HC: HostComponent> Copy for HostComponentDataHandle<HC> {}
+
+impl<HC: HostComponent> Clone for HostComponentDataHandle<HC> {
+    fn clone(&self) -> Self {
+        Self {
+            idx: self.idx,
+            _phantom: PhantomData,
+        }
+    }
+}
+
+type DataBuilder = Box<dyn Fn() -> Box<dyn Any + Send> + Send + Sync>;
+
+pub struct HostComponentsBuilder {
+    data_builders: Vec<DataBuilder>,
+}
+
+impl HostComponentsBuilder {
+    pub fn add_host_component<T: Send, HC: HostComponent + 'static>(
+        &mut self,
+        linker: &mut Linker<T>,
+        host_component: HC,
+    ) -> Result<HostComponentDataHandle<HC>> {
+        let idx = self.data_builders.len();
+        self.data_builders
+            .push(Box::new(move || Box::new(host_component.build_data())));
+        HC::add_to_linker(linker, move |data| {
+            data.host_components_data
+                .get_or_insert_idx(idx)
+                .downcast_mut()
+                .unwrap()
+        })?;
+        Ok(HostComponentDataHandle::<HC> {
+            idx,
+            _phantom: PhantomData,
+        })
+    }
+
+    pub fn build(self) -> HostComponents {
+        let data_builders = Arc::new(self.data_builders);
+        HostComponents { data_builders }
+    }
+}
+
+pub struct HostComponents {
+    data_builders: Arc<Vec<DataBuilder>>,
+}
+
+impl HostComponents {
+    pub fn builder() -> HostComponentsBuilder {
+        HostComponentsBuilder {
+            data_builders: Default::default(),
+        }
+    }
+
+    pub fn new_data(&self) -> HostComponentsData {
+        // Fill with `None`
+        let data = std::iter::repeat_with(Default::default)
+            .take(self.data_builders.len())
+            .collect();
+        HostComponentsData {
+            data,
+            data_builders: self.data_builders.clone(),
+        }
+    }
+}
+
+/// Holds a heterogenous set of [`HostComponent::Data`]s.
+pub struct HostComponentsData {
+    data: Vec<Option<Box<dyn Any + Send>>>,
+    data_builders: Arc<Vec<DataBuilder>>,
+}
+
+impl HostComponentsData {
+    /// Sets the [`HostComponent::Data`] for the given `handle`.
+    pub fn set<HC: HostComponent>(&mut self, handle: HostComponentDataHandle<HC>, data: HC::Data) {
+        self.data[handle.idx] = Some(Box::new(data));
+    }
+
+    /// Retrieves a mutable reference to [`HostComponent::Data`] for the given `handle`.
+    ///
+    /// If unset, the data will be initialized with [`HostComponent::build_data`].
+    pub fn get_or_insert<HC: HostComponent>(
+        &mut self,
+        handle: HostComponentDataHandle<HC>,
+    ) -> &mut HC::Data {
+        let x = self.get_or_insert_idx(handle.idx);
+        x.downcast_mut().unwrap()
+    }
+
+    fn get_or_insert_idx(&mut self, idx: usize) -> &mut Box<dyn Any + Send> {
+        self.data[idx].get_or_insert_with(|| self.data_builders[idx]())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    struct TestHC;
+
+    impl HostComponent for TestHC {
```

Comment:
If you are just trying to test code flow here, you could use mockall instead of creating a test struct: https://docs.rs/mockall/latest/mockall/

Reply:
This would be good to discuss with the team. Mocking isn't _super_ common in the Rust ecosystem and there is always some learning overhead in new frameworks like that, but it would certainly be helpful in some cases like this.